app/assets/src/components/views/PipelineViz.jsx

```diff
@@ -0,0 +1,379 @@
+import React from "react";
+import { forEach } from "lodash";
+import PropTypes from "prop-types";
+import ReactPanZoom from "@ajainarayanan/react-pan-zoom";
+
+import RemoveIcon from "~/components/ui/icons/RemoveIcon";
+import NetworkGraph from "~/components/visualizations/NetworkGraph.js";
+import cs from "./pipeline_viz.scss";
+
+const START_NODE_ID = -1;
+const END_NODE_ID = -2;
+
+class PipelineViz extends React.Component {
```

Comment:
Have you considered a function component with hooks? It looks like it could handle all of this, and be a little cleaner, though I don't want to assume.

Reply:
Just curious @dcwither , what observations triggered you to make that suggestion?

Reply:
- use of a `componentDidMount` when a `componentDidUpdate` is probably also necessary (solved by dependencies in the `useEffect` hook). 
- multiple pieces of state that are modified independently, also some that could benefit from useReducer
- I think it would also help identify things that are being mutated that shouldn't be

Reply:
Good points! 
However this is just the first pass on this component with basic functionality, so I am not sure we want to go with a functional component. But we will definitely keep that in mind.